회원가입 기초기능 구현

chanwoothecali opened the pull request:
## 회원가입 기초기능 구현
- User의 가장 기초가 되는 회원가입 기능을 구현
- 타 기능을 개발하기 전 User 관련 틀이 잡혀있어야 하기 때문에 가장 기초적인 기능만을 추가
- 추가적인 기능은 추후 PR 예정
### DB test 추가
- UserRepositoryTest는 DB 접속만을 위한 테스트이며 타 도메인은 repositoryTest 생략
### spring-restdocs-mockmvc
- restdocs를 이용한 ControllerTest 추가. 추후 api docs 설정도 추가 예정

chanwoothecali committed:
유저 엔티티 및 레포지토리 추가

* 유저 엔티티 추가
    - user 예약어로 인해 테이블명 users로 지정
* 유저 레포지토리 추가 및 테스트 추가
    - db 접속 테스트를 위한 테스트 레포지토리

chanwoothecali committed:
유저 엔티티 생성/수정시간 칼럼 관련 수정

- BaseEntity 추가 예정이므로 생성/수정시간 칼럼 생성자에서 제외
- 추후 JPA Auditing 활성화 예정이므로 nullable false 속성 주석처리
- 그로 인한 기존 UserRepository 테스트 수정

chanwoothecali committed:
회원가입 및 이메일 중복 가입 예외 처리

* UserService
    - 회원가입 로직 추가
* UserDto
    - 회원가입 시 정보를 받을 SaveRequest DTO 추가
    - User 관련 DTO를 UserDTO로 관리
* 회원가입시 이메일 중복 예외 처리

chanwoothecali committed:
UserController 회원가입 추가

* UserController
    - 회원가입 기능만 추가
    - restdocs를 이용한 테스트 추가
* 회원가입시 User 기본 역할 부여

f-lab-michael commented on src/main/java/com/flab/modu/users/controller/UserController.java:
```diff
@@ -0,0 +1,24 @@
+
+    private final UserService userService;
+
+    @PostMapping("/user")
```
url 에 들어가는 resource 는 가능하면 복수로 하는것이 좋습니다.

f-lab-michael commented on src/main/java/com/flab/modu/users/controller/UserController.java:
```diff
@@ -0,0 +1,24 @@
+    private final UserService userService;
+
+    @PostMapping("/user")
+    public ResponseEntity<Void> createUser(@RequestBody @Valid SaveRequest saveRequest) {
```
SaveRequest 라고 표현시 어떤걸 save 하는지 알기 어려울것 같습니다. User 라는 단어를 추가하는게 어떨까요? CreateUserRequest 등으로 사용할수있을것 같네요

chanwoothecali replied:
DTO 클래스는 패키지에 생성하는 것보다 관련있는 것들끼리 묶기위해 하나의 DTO 내에 static class로 선언해두었습니다.
SaveRequest 클래스도 UserDto 내부에 static class로 선언해둔 것이라 중복명칭을 방지하기 위해 User를 생략했습니다!
UserDto.SaveRequest로 표현하는 편이 나을까요? 아니면 UserDto 내부에 두되 CreateUserRequest로 명칭을 바꾸는 게 나을까요?

f-lab-michael replied:
두가지 방식중에 마음에 드시는 방식으로 하시면 될것 같습니다. 
그리고 save 라는 단어보다는 create 라는 단어가 더 명시적일것 같습니다. 
수정도 save 라고 볼수있을것 같아서요

f-lab-michael commented on src/main/java/com/flab/modu/users/domain/common/UserRole.java:
```diff
@@ -0,0 +1,5 @@
+package com.flab.modu.users.domain.common;
+
+public enum UserRole {
+    NORMAL, SELLER, ADMIN;
```
NORMAL 대신 좀더 의미를 담은 단어인 MEMBER나 BUYER 등은 어떨까요?

chanwoothecali replied:
도메인 특성 상 BUYER가 맞는 것 같습니다. BUYER로 변경해두겠습니다.

f-lab-michael commented on src/main/java/com/flab/modu/users/domain/entity/User.java:
```diff
@@ -0,0 +1,56 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false, unique = true)
```
어노테이션으로 유니크 설정을 하는경우 단점은 없을까요?

chanwoothecali replied:
jpa로 데이터베이스를 생성하는 경우에만 적용됩니다.
로컬 환경에서는 ddl-auto 속성이 create라 unique 설정이 가능하지만, 추후 다른 환경에서는 직접 설정해주어야 합니다.

참고: 
https://stackoverflow.com/questions/30460596/jpa-column-unique-true-what-is-really-point-of-having-unique-attribute
https://velog.io/@joshuara7235/ISSUEs-JPA%EB%A1%9C-%EC%83%9D%EC%84%B1%ED%95%9C-entity-column%EC%97%90%EC%84%9C-unique-%EC%A0%9C%EC%95%BD%EC%A1%B0%EA%B1%B4-%EC%82%AD%EC%A0%9C%ED%95%98%EA%B8%B0

f-lab-michael replied:
네 추가로 컬럼으로 설정하는경우 유니크 인덱스의 이름이 자동으로 설정되어 가독성이 떨어지는 단점도 존재합니다.

f-lab-michael commented on src/main/java/com/flab/modu/users/domain/entity/User.java:
```diff
@@ -0,0 +1,56 @@
+
+    private String phoneNumber;
+
+//    @Column(nullable = false, updatable = false)
```
주석처리가 되어 있습니다.

f-lab-michael commented on src/main/java/com/flab/modu/users/domain/entity/User.java:
```diff
@@ -0,0 +1,56 @@
+//    @Column(nullable = false, updatable = false)
+    private LocalDateTime createdAt;
+
+//    @Column(nullable = false)
```
여기도

f-lab-michael commented on src/main/java/com/flab/modu/users/service/UserService.java:
```diff
@@ -0,0 +1,27 @@
+
+    private final UserRepository userRepository;
+
+    public void createUser(SaveRequest saveRequest) {
```
받아온 암호 정보를 그래도 디비에 저장하게 되면 어떤 문제가 생길까요?

chanwoothecali replied:
데이터베이스에 접근 가능한 모든 사람에게 암호가 노출된다는 문제가 있습니다.
해당 PR 이후 다음 PR에서 바로 수정하도록 하겠습니다.

f-lab-michael commented on src/test/java/com/flab/modu/users/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,68 @@
+
+    @MockBean
+    private UserService userService;
+    @Autowired
```
뉴라인이 빠졌네요

f-lab-michael commented on src/test/java/com/flab/modu/users/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,68 @@
+@ExtendWith(RestDocumentationExtension.class)
+@AutoConfigureRestDocs
+@WebMvcTest(UserController.class)
+@MockBean(JpaMetamodelMappingContext.class)
```
이 부분은 왜 들어갔는지 설명해주실수있을까요?

chanwoothecali replied:
JpaMetamodelMappingContext.class는 JPA 관련 Bean이 컨트롤러 테스트에 필요할 때 사용하는 class입니다.
@EnableJpaAuditing을 사용할 때 필요한 기능이라 현재 시점에는 필요없는 것 같습니다. 삭제하도록 하겠습니다.

f-lab-michael commented on src/test/java/com/flab/modu/users/service/UserServiceTest.java:
```diff
@@ -0,0 +1,55 @@
+    public void emailDuplicate() {
+        // given
+        SaveRequest testUserData = createTestUserData();
+        userService.createUser(testUserData);
```
테스트 코드를 실행하는 부분이  when 이고

chanwoothecali replied:
테스트 작성에만 급급했던 것 같습니다. Mokito와 BDDMokito에 대해 공부해보고 수정하도록 하겠습니다.

f-lab-michael commented on src/test/java/com/flab/modu/users/service/UserServiceTest.java:
```diff
@@ -0,0 +1,55 @@
+        String existingEmail = "test@modu.com";
+
+        // when
+        when(userRepository.existsByEmail(existingEmail)).thenReturn(true);
```
조건을 설정하는 부분이 given 영역입니다.
사용하신 함수라 when 이여서 헤깔리시는것 같은데 MockitoBDD에서 제공하는 given 함수를 쓰시면 동일한 모킹을 할수있습니다.

f-lab-michael commented on src/test/java/com/flab/modu/users/service/UserServiceTest.java:
```diff
@@ -0,0 +1,55 @@
+
+        // then
+        assertThrows(DuplicatedEmailException.class, () -> userService.createUser(saveRequest));
+        verify(userRepository, atLeastOnce()).existsByEmail(existingEmail);
```
bdd 로 하실꺼면 함수 호출 검증도 then 함수를 쓰시면 됩니다.

chanwoothecali committed:
PR 피드백 반영

# UserRole
  - NOMARL 대신에 BUYER로 변경
# UserDto
  - SaveRequest에서 CreateRequest 변경
  - CreateRequest를 사용하는 모든 곳에 UserDto 명시
# UserController
  - url에 user 리소스명 복수화 (users)
# UserControllerTest
  - 필요없는 애노테이션 삭제
# UserServiceTest
  - BDDMockito 사용
# User
  - JPA 애노테이션을 이용한 unique 부여 삭제
  - 주석 커밋한 부분 삭제

f-lab-michael commented on src/test/java/com/flab/modu/users/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,67 @@
+                fieldWithPath("name").type(JsonFieldType.STRING)
+                    .description("name"),
+                fieldWithPath("phoneNumber").type(JsonFieldType.STRING).description("phoneNumber")
+            )));
```
실제 userService 의 createUser 를 올바른 파라미터로 호출했는지도 테스트 해주세요

f-lab-michael commented on src/test/java/com/flab/modu/users/service/UserServiceTest.java:
```diff
@@ -0,0 +1,47 @@
+        // then
+        then(userRepository).should().existsByEmail("test@modu.com");
+    }
+}
```
모든 파일의 마지막은 뉴라인으로 끝내느게 좋습니다

f-lab-michael commented on src/test/java/com/flab/modu/users/service/UserServiceTest.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @Test
+    @DisplayName("이메일 중복은 회원가입에 실패한다.")
```
정상케이스도 테스트 해주세요

f-lab-michael commented on src/test/java/com/flab/modu/users/service/UserServiceTest.java:
```diff
@@ -0,0 +1,47 @@
+        assertThrows(DuplicatedEmailException.class, () -> userService.createUser(createRequest));
+
+        // then
+        then(userRepository).should().existsByEmail("test@modu.com");
```
이후에 save 를 호출하지 않는것도 테스트가 가능합니다

chanwoothecali committed:
예외처리 컨트롤러 추가

# ExceptionController
    - RestControllerAdvice & ExceptionHandler를 이용해 컨트롤러 단의 Exception을 공통으로 처리

chanwoothecali committed:
2차 피드백 반영 및 dto 내 유효성 검증 변경

# UserControllerTest
    - 이메일 중복 시 실패 테스트 추가
# UserServiceTest
    - 정상 가입 테스트 추가
    - 실패 시 실제로 다음 단계기 이뤄지지 않았는지에 대한 검증 추가
# UserDto
    - Request Parameter 검증 방식 변경

chanwoothecali committed:
각 소스마다 EOL 추가

eol이 없는 소스마다 new line을 추가

f-lab-michael commented on src/test/java/com/flab/modu/users/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,104 @@
+            .phoneNumber("01012345678")
+            .build();
+
+        doNothing().when(userService).createUser(createRequest);
```
이것도 given 으로 통일하면 어떨까요?

f-lab-michael commented on src/test/java/com/flab/modu/users/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,104 @@
+                    fieldWithPath("phoneNumber").type(JsonFieldType.STRING)
+                        .description("휴대폰 번호")
+                )
+            ));
```
then 절을 통해 userService를 올바른 파라미터로 호출했는지 여부를 검증할수있습니다.

f-lab-michael commented on src/test/java/com/flab/modu/users/service/UserServiceTest.java:
```diff
@@ -0,0 +1,82 @@
+        // then
+        then(userRepository).should().existsByEmail("test@modu.com");
+        verify(userRepository, atLeastOnce()).existsByEmail(existingEmail);
+        verify(userRepository, never()).save(any());
```
then 절을 통해서도 할수있습니다.
`then(userRepository).shouldHaveNoInteractions();`

f-lab-michael commented on src/test/java/com/flab/modu/users/service/UserServiceTest.java:
```diff
@@ -0,0 +1,82 @@
+
+        // then
+        then(userRepository).should().existsByEmail("test@modu.com");
+        verify(userRepository, atLeastOnce()).existsByEmail(existingEmail);
```
이 내용은 위의 then 절과 동일한 일을 하는것 같습니다.

chanwoothecali committed:
Test 코드 리팩토링

- Mockito 대신 BDDMockito를 사용하도록 수정

chanwoothecali committed:
UserRepository 불필요 접근제한자 제거

Repository 메소드에 불필요한 public 접근 제한자를 삭제
